runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,118 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
+            // the type of the object at dest, so all of the above cases will be treated like case (1).
+
+            //if offheap is not enabled, we will never need to load dataAddr. So we know for sure that we can calculate the final
+            //destination address right away and only pass in three children to setmemoryEvaluator() (dest, len, byteValue)
+
+            //if offheap is not enabled, default to case (1):
+            // - do NOT to load dataAddr to calculate final destination address
+            // - calculate the final destination address right away and only pass three children into setmemoryEvaluator (dest, len, byteValue)
+            bool loadDataAddr = false;
+            bool separateDestAndOffset = false;
+
+         #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+            if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit())
+               {
+               if (dest->getSymbolReference() != NULL)
+                  {
+                  //check dstBaseAddrNode type at compile time
+                  int length;
+                  const char *objTypeSig = dest->getSymbolReference()->getTypeSignature(length);
+
+                  // There are four cases where we cannot be sure of the Object type at compile time:
+                  // 1.) The object's type signature is unavailable/unknown
+                  // 2.) The signature names java/lang/Object
+                  // 3.) The signature names some interface type, wich is a problem because interface types are
+                  //     not checked by the verifier
+                  // 4.) The signature belongs to a parameter of the calling method, which is a problem because
+                  //     parameters can share slots with variables of other, potentially incompatible types
+                  bool objTypeUnknown;
+                  TR_OpaqueClassBlock *objClass = NULL;
+
+                  if (objTypeSig == NULL)
+                     {
+                     objTypeUnknown = true;
+                     }
+                  else
+                     {
+                     objClass = fej9->getClassFromSignature(objTypeSig, length, dest->getSymbolReference()->getOwningMethod(comp));
+                     bool isParameter = dest->getSymbolReference()->getSymbol()->isParm();
+
+                     objTypeUnknown = objClass == NULL || objClass == comp->getObjectClassPointer() ||
+                                      TR::Compiler->cls.isInterfaceClass(comp, objClass) || isParameter;
+                     }
+
+                  //case (2): want to load dataAddr and calculate final dest address right away if both of the following conditions are met:
+                  //          1.) Object is known to be an array at compile time
+                  //          2.) Object is known to be a non-NULL reference at compile time
+                  loadDataAddr = !objTypeUnknown && objTypeSig[0] == '[' && dest->isNonNull();
```

Comment:
so, what are you going to do, `if dest->isNonNull() is false`? you simply skip loading dataAddr but at the same time making it 3-children? or, somehow you think that case is impossible (i.e. compiler doesn't know if a variable is NULL at runtime), let's assume it is a known array type. I strongly recommend you think through all the cases.

nitpick:  coding style -- bracket the condition clauses necessarily like:  `(!objTypeUnknown) && (objTypeSig[0] == '[') && dest->isNonNull()`

Reply:
Ok, I modified the conditions to account for all possible combinations of object types/NULL statuses (and tested accordingly as well). Notably, since knowing that the object is a NULL reference at compile time immediately tells us that we do not want to load dataAddr, I moved that condition to the outermost if statement (where we check if offheap is enabled), which makes the evaluation of `loadDataAddr` and `separateDestAndOffset` much simpler and easier to read: 
```
if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit() && (!dest->isNull()))
{
.
.
.
//case (2): want to load dataAddr and calculate final dest address right away if both of the following conditions are met:
//          1.) Object is known to be an array at compile time
//          2.) Object is known to be a non-NULL reference at compile time
loadDataAddr = (!objTypeUnknown) && (objTypeSig[0] == '[') && dest->isNonNull();

//case (3): want to keep dest and destOffset separate so that we can perform runtime NULL/array tests if at least one oft he following conditions are met:
//          1.) Object type is unknown at compile time
//          2.) Object is known to be an array at compile time AND 
//              it is NOT known that the object is a non-NULL reference at compile time
separateDestAndOffset = objTypeUnknown || ((objTypeSig[0] == '[') && (!dest->isNonNull()));
```
https://github.com/midronij/openj9/blob/fbd5fc5ca1f441452d3539a7c23d9d7c24d1d795/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L11686-L11738